### PR TITLE
handle set_buffer_scale and output_scale

### PR DESCRIPTION
**Pull request by ss1h2a3tw**
Add set output scale, get output scale,get output scaled size api
with @david50407 's help
for fixing https://github.com/Cloudef/wlc/issues/57
I'm not sure about whether the pointer handling is right, I don't know how to check the recursive part.
Wish to make some progress to HiDPI support XD.

### CODE REVIEW

File: example/example.c [hunk 1]
```diff
@@ -4,6 +4,8 @@
 #include <chck/math/math.h>
 #include <linux/input.h>
 
+#define SCALE 2
```
Comment (jserv) [inline example/example.c]: Why did you hardcode `2` here?

File: example/example.c [hunk 2]
```diff
@@ -254,7 +258,6 @@ int
 main(void)
 {
    wlc_log_set_handler(cb_log);
-
```
Comment (jserv) [inline example/example.c]: Don't change the original coding style.

File: src/compositor/output.c [hunk 3]
```diff
@@ -868,12 +868,52 @@ wlc_output_get_resolution(wlc_handle output)
...
 }
 
+bool
+wlc_output_set_scale_ptr(struct wlc_output *output, int32_t scale)
+{
+   if (!output)
+      return false;
+
+   assert(scale);
+   wlc_log(WLC_LOG_WARN, "Set scale of output %" PRIuWLC "to %u", convert_to_wlc_handle(output), scale);
```
Comment (jserv) [inline src/compositor/output.c]: There is no flags following character `%` in the format string.

File: src/compositor/output.c [hunk 4]
```diff
@@ -868,12 +868,52 @@ wlc_output_get_resolution(wlc_handle output)
...
 wlc_output_set_resolution(wlc_handle output, const struct wlc_size *resolution)
 {
    wlc_output_set_resolution_ptr(convert_from_wlc_handle(output, "output"), resolution);
 }
 
+WLC_API void
+wlc_output_set_scale(wlc_handle output, int32_t scale)
+{
+   wlc_output_set_scale_ptr(convert_from_wlc_handle(output, "output"), scale);
+}
```
Comment (jserv) [inline src/compositor/output.c]: Append the space line to separate each function.

File: src/compositor/seat/pointer.c [hunk 5]
```diff
@@ -214,6 +214,9 @@ focus_view(struct wlc_pointer *pointer, struct wlc_surface *surf, wlc_handle old
    if (!surf || !(surface = convert_to_wl_resource(surf, "surface")))
       return;
 
+   uint32_t scale = surf->pending.scale;
+   if(scale==0)scale=1;
```
Comment (jserv) [inline src/compositor/seat/pointer.c]: Follow the consistent coding style.

File: src/platform/render/gles2.c [hunk 6]
```diff
@@ -734,6 +734,19 @@ view_paint(struct ctx *context, struct wlc_view *view)
 
    struct wlc_geometry geometry;
    wlc_view_get_bounds(view, &geometry, &settings.visible);
+   wlc_handle output = wlc_view_get_output(convert_to_wlc_handle(view));
+
+   int32_t scale=wlc_output_get_scale(output);
+   //wlc_log(WLC_LOG_WARN, "origin x %d y %d size x %u y %u  scale %u",geometry.origin.x,geometry.origin.y,geometry.size.w,geometry.size.h,scale);
```
Comment (jserv) [inline src/platform/render/gles2.c]: Don't leave the dead code.

File: src/platform/render/gles2.c [hunk 7]
```diff
@@ -734,6 +734,19 @@ view_paint(struct ctx *context, struct wlc_view *view)
...
+   geometry.origin.x*=scale;
```
Comment (jserv) [inline src/platform/render/gles2.c]: Follow the consistent coding style.

File: include/wlc/wlc.h [hunk 8]
```diff
@@ -284,6 +284,15 @@ const struct wlc_size* wlc_output_get_resolution(wlc_handle output);
...
 WLC_NONULL void wlc_output_set_resolution(wlc_handle output, const struct wlc_size *resolution);
 
+/** Get scale factor. */
+int32_t wlc_output_get_scale(wlc_handle output);
+
+/** Set scale factor. */
+void wlc_output_set_scale(wlc_handle output, int32_t scale);
+
+/** Get scaled size. */
+bool wlc_output_get_scaled_size(wlc_handle output, struct wlc_size *out_size);
```
Comment (jserv) [inline include/wlc/wlc.h]: Is function `wlc_output_get_scaled_size` necessary? Can we simply calculate demanding size each time?

File: example/example.c (hunk 1 above)
Comment (ss1h2a3tw) [inline example/example.c]: This is a example of a compositor working on scale 2.

Comment (jserv) [inline example/example.c]: You should specify it to reflect its real functionality.

File: include/wlc/wlc.h (hunk 8 above)
Comment (ss1h2a3tw) [inline include/wlc/wlc.h]: It is not necessary, but useful. Because developer may meet the situation that there are multiple monitors with different scale. With this function, developer can directly get the global coordinate of the compositor.

File: src/platform/render/gles2.c (hunk 6 above)
Comment (ss1h2a3tw) [inline src/platform/render/gles2.c]: Sorry, I miss this line when I cleaning it up.

File: example/example.c (hunk 1 above)
Comment (ss1h2a3tw) [inline example/example.c]: OK

**Comment by jserv** [comment]
@ss1h2a3tw : You should squash the coding style fixes into single commit via `git rebase -i`.

File: include/wlc/wlc.h (hunk 8 above)
Comment (david50407) [inline include/wlc/wlc.h]: @jserv Yes, it's not necessary, but if we considered that compositor use the global coordinate which's dimension is same as the surface coordinate. 
That would be useful that developer of the compositor can just get the result from wlc without handle this my themselves.

File: src/compositor/output.c [hunk 9]
```diff
@@ -868,12 +868,52 @@ wlc_output_get_resolution(wlc_handle output)
...
+WLC_API int32_t
+wlc_output_get_scale(wlc_handle output)
+{
+   struct wlc_output *o = convert_from_wlc_handle(output, "output");
+   assert(o);
+   return o->information.scale;
+}
+
+WLC_API bool
```
Comment (jserv) [inline src/compositor/output.c]: Why does it return boolean here?

Comment (ss1h2a3tw) [inline src/compositor/output.c]: @david50407

Comment (david50407) [inline src/compositor/output.c]: @jserv check if the operation is succeed. Fails may caused by NULL out param or cannot get the resolution of the output (may a invalid output).

Comment (jserv) [inline src/compositor/output.c]: I think it would be straightforward if a compositor takes care of scaling information and makes precise decision itself.

Comment (david50407) [inline src/compositor/output.c]: wlc is for helping to building compositor, wlc could be a useful interface layer for developers to make a compositor easily and quickly, but if we wants to get the original resolution, we still can use `wlc_output_get_resolution`.
This is just a helper function.

Comment (david50407) [inline src/compositor/output.c]: And for some scaling tricks, we wish that wlc can handle non-integer scaling, and the fake coordinate will be calculated in wlc, so that developers can make the compositor in the global  coordinate world directly.

File: src/compositor/output.c (hunk 4 above)
Comment (Cloudef) [inline src/compositor/output.c]: Hmm doesn't it make sense to have output_set_scale in public API? Normally I would assume the backend handles the scales.

File: include/wlc/wlc.h (hunk 8 above)
Comment (Cloudef) [inline include/wlc/wlc.h]: Right now wlc doesn't have concept of global coordinates (and may not have, as there is no standard for monitor placement)

File: src/compositor/output.c (hunk 9 above)
Comment (Cloudef) [inline src/compositor/output.c]: non-integral scaling can be achieved by combination of scale and wlc_output_set_resolution. Non integral scaling for surfaces is not supported by wayland protocol. Also supporting that would be really PITA when you consider buffer sizes could be non integral. Thus it's better to set nearest integral scale to output and then compensate with rendering resolution (not neccessarily modeline).

File: src/compositor/seat/pointer.c [hunk 10]
```diff
@@ -139,10 +139,10 @@ pointer_paint(struct wlc_pointer *pointer, struct wlc_output *output)
 
    if (!pointer || output != active_output(pointer))
       return;
-
+   const int32_t scale=output->information.scale;
```
Comment (Cloudef) [inline src/compositor/seat/pointer.c]: whitespace around operators.

File: src/compositor/output.c [hunk 11]
```diff
@@ -874,6 +874,47 @@ wlc_output_set_resolution(wlc_handle output, const struct wlc_size *resolution)
...
+wlc_output_get_scaled_size(wlc_handle output, struct wlc_size *out_size)
+{
+   if (!out_size)
```
Comment (Cloudef) [inline src/compositor/output.c]: assert instead, as out_size is not optional parameter at all to this. Though I'm questioning the usefulness of this function. It also might make sense to zero initialize out_size. It would make possible to use this function then regardless of return value as long as zero size is fine.

But for now, I think this function can be removed as I don't think of it as neccessity, writing wrapper function should be straight forward.

File: src/compositor/seat/pointer.c [hunk 12]
```diff
@@ -139,10 +139,10 @@ pointer_paint(struct wlc_pointer *pointer, struct wlc_output *output)
...
    const struct wlc_point pos = {
-      chck_clamp(pointer->pos.x, 0, output->resolution.w),
-      chck_clamp(pointer->pos.y, 0, output->resolution.h)
+      chck_clamp(pointer->pos.x*scale, 0, output->resolution.w),
+      chck_clamp(pointer->pos.y*scale, 0, output->resolution.h)
```
Comment (Cloudef) [inline src/compositor/seat/pointer.c]: Whitespace around operators.

File: src/compositor/seat/pointer.c [hunk 13]
```diff
@@ -225,7 +227,7 @@ focus_view(struct wlc_pointer *pointer, struct wlc_surface *surf, wlc_handle old
          wlc_log(WLC_LOG_WARN, "Failed to push focused pointer resource to pool (out of memory?)");
 
       uint32_t serial = wl_display_next_serial(wlc_display());
-      wl_pointer_send_enter(wr, serial, surface, wl_fixed_from_double(pos->x), wl_fixed_from_double(pos->y));
+      wl_pointer_send_enter(wr, serial, surface, wl_fixed_from_double(pos->x/scale), wl_fixed_from_double(pos->y/scale));
```
Comment (Cloudef) [inline src/compositor/seat/pointer.c]: I will skip the view.c and pointer.c code changes and detail my thoughts about how this functionality could be achieved with less changes in separate comment.

File: src/compositor/view.c [hunk 14]
```diff
@@ -205,8 +207,8 @@ wlc_view_get_bounds(struct wlc_view *view, struct wlc_geometry *out_bounds, stru
       // xdg-surface client that draws drop shadows or other stuff.
       out_bounds->origin.x -= view->surface_commit.visible.origin.x;
       out_bounds->origin.y -= view->surface_commit.visible.origin.y;
-      out_bounds->size.w += surface->size.w - view->surface_commit.visible.size.w;
-      out_bounds->size.h += surface->size.h - view->surface_commit.visible.size.h;
+      out_bounds->size.w += surface->size.w / scale - view->surface_commit.visible.size.w;
+      out_bounds->size.h += surface->size.h / scale - view->surface_commit.visible.size.h;
```
Comment (Cloudef) [inline src/compositor/view.c]: Why is this necessary ?

File: src/compositor/view.c [hunk 15]
```diff
@@ -215,9 +217,13 @@ wlc_view_get_bounds(struct wlc_view *view, struct wlc_geometry *out_bounds, stru
    if (!out_visible)
       return;
 
+   struct wlc_size temp=surface->size;
+   temp.w /= scale;
+   temp.h /= scale;
```
Comment (Cloudef) [inline src/compositor/view.c]: Ditto?

File: src/platform/render/gles2.c [hunk 16]
```diff
@@ -734,6 +734,18 @@ view_paint(struct ctx *context, struct wlc_view *view)
...
+   wlc_handle output = wlc_view_get_output(convert_to_wlc_handle(view));
+   int32_t scale=wlc_output_get_scale(output);
+   geometry.origin.x*=scale;
+   geometry.origin.y*=scale;
+   geometry.size.w*=scale;
+   geometry.size.h*=scale;
+   settings.visible.origin.x*=scale;
+   settings.visible.origin.y*=scale;
+   settings.visible.size.w*=scale;
+   settings.visible.size.h*=scale;
```
Comment (Cloudef) [inline src/platform/render/gles2.c]: Whitespaces around operations.

File: src/resources/types/surface.c [hunk 17]
```diff
@@ -54,6 +54,9 @@ commit_state(struct wlc_surface *surface, struct wlc_surface_state *pending, str
    state_set_buffer(out, convert_from_wlc_resource(pending->buffer, "buffer"));
    state_set_buffer(pending, NULL);
 
+   out->scale = pending->scale;
```
Comment (Cloudef) [inline src/resources/types/surface.c]: out->scale = chck_max32(pending->scale, 1);

File: src/resources/types/surface.c [hunk 18]
```diff
@@ -391,6 +394,7 @@ wlc_surface(struct wlc_surface *surface)
        !chck_iter_pool(&surface->subsurface_list, 4, 0, sizeof(wlc_resource)))
       goto fail;
 
+   surface->commit.scale = 1;
```
Comment (Cloudef) [inline src/resources/types/surface.c]: Is this neccessary?

File: src/compositor/seat/pointer.c (hunk 13 above)
Comment (Cloudef) [inline src/compositor/seat/pointer.c]: whietspace around operations.

File: src/compositor/seat/pointer.c [hunk 19]
```diff
@@ -344,7 +350,7 @@ wlc_pointer_motion(struct wlc_pointer *pointer, uint32_t time, bool pass)
       if (!(wr = wl_resource_from_wlc_resource(*r, "pointer")))
          continue;
 
-      wl_pointer_send_motion(wr, time, wl_fixed_from_double(d.x), wl_fixed_from_double(d.y));
+      wl_pointer_send_motion(wr, time, wl_fixed_from_double(d.x/scale), wl_fixed_from_double(d.y/scale));
```
Comment (Cloudef) [inline src/compositor/seat/pointer.c]: whitespace around oeprators.

File: src/compositor/seat/seat.c [hunk 20]
```diff
@@ -145,14 +145,15 @@ input_event(struct wl_listener *listener, void *data)
...
+   const int32_t scale=output->information.scale;
    switch (ev->type) {
       case WLC_INPUT_EVENT_MOTION:
       {
          const struct wlc_size resolution = (output ? output->resolution : wlc_size_zero);
 
          const struct wlc_pointer_origin pos = {
-            chck_clamp(seat->pointer.pos.x + ev->motion.dx, 0, resolution.w),
-            chck_clamp(seat->pointer.pos.y + ev->motion.dy, 0, resolution.h),
+            chck_clamp(seat->pointer.pos.x + ev->motion.dx, 0, resolution.w/scale),
```
Comment (Cloudef) [inline src/compositor/seat/seat.c]: I would imagine the motion would need to be divided not the boundary.

File: src/compositor/seat/seat.c [hunk 21]
```diff
@@ -165,8 +166,8 @@ input_event(struct wl_listener *listener, void *data)
          const struct wlc_size resolution = (output ? output->resolution : wlc_size_zero);
 
          const struct wlc_pointer_origin pos = {
-            ev->motion_abs.x(ev->motion_abs.internal, resolution.w),
-            ev->motion_abs.y(ev->motion_abs.internal, resolution.h)
+            ev->motion_abs.x(ev->motion_abs.internal, resolution.w/scale),
+            ev->motion_abs.y(ev->motion_abs.internal, resolution.h/scale)
```
Comment (Cloudef) [inline src/compositor/seat/seat.c]: Ditto in the cases here and below.

**Comment by Cloudef** [comment]
Does the client expect coordinates it receives to be as if they were on 1x scale. That is we need to map for example 2x, mouse movement over surface to 1x scale coordinate system?       

Looking at this PR, it may be enough to simply just handle scaling in the wlc_view_get_bounds and wlc_view_get_visible_bounds function, and everything else will work as normal (hit testing, culling, rendering, input..) As input already does coordinate scaling to _surface size_ which is the 1x size without scaling. Can you try this and report that other changes are not needed?

For other things, I mentioned above wlc_output_set_scale sounds weird, but perhaps it can be useful to set custom scale on output. However does EDID / drm / etc.. give this information for us? Or does weston simply just use scale parameter in ini configuration as well.

File: src/resources/types/surface.c (hunk 18 above)
Comment (ss1h2a3tw) [inline src/resources/types/surface.c]: Because commit scale is zero at the begining , it will cause dividing by zero.

**Comment by ss1h2a3tw** [comment]
I will test it later, thanks for your reply.

File: src/resources/types/surface.c (hunk 18 above)
Comment (Cloudef) [inline src/resources/types/surface.c]: Fair enough. It sounds like 0 scale for initial doesn't make much sense anyways. We consider and check client set surface scale at <= 0 illegal anyways.

**Comment by ss1h2a3tw** [comment]
As only modifying get_bound result in the cursor's position differ between rendered on screen and inside client.
img_20160622_220356

in the photo , the cursor is in the title bar of weston-terminal (right middle in "progs"),
but the play button in the rhythmbox is highlighted.

using the code below
https://github.com/Cloudef/wlc/compare/master...ss1h2a3tw:test

File: src/compositor/view.c (hunk 14 above)
Comment (ss1h2a3tw) [inline src/compositor/view.c]: surface's size seems to be the geometry*buffer_scale , but the bounds' coordinate should be  same as geometry?

File: src/compositor/view.c (hunk 15 above)
Comment (ss1h2a3tw) [inline src/compositor/view.c]: I'm trying to just memcpy visible part when scaling the surface.

**Comment by ss1h2a3tw** [comment]
@Cloudef 
Maybe changing line number 290 of surface.c can help?
It seems like it is directly copy the size of buffer, which explains why I need to divide surface size by scale.

**Comment by Cloudef** [comment]
Dividing surface's size is enough for eveything to work? That sounds like good idea to me if so. Do we actually need the 2x size information anywhere now?

**Comment by ss1h2a3tw** [comment]
Yeah, looks like buffer_scale can be handled as dividing surface's size.
supporting output scale will need to change seat/pointer.c , gles2.c

**Comment by Cloudef** [comment]
Ok, so the steps we have done here.

1) Divide surface size with its scale (1 == unaffected, 2 == half, so on..)
2) Multiply input coordinates with output's scale, divide boundary by output scale. 
3) Render surfaces by multiplying surface geometry with output scale. (1 == unaffected, 2 == multiple of size, so on..)

1 and 3 seems like good idea as it avoids conditional calculation depending if surface size is less or more than output scale and so on.

What's the purpose of step 2? How does this work with scale 1 surfaces on scale 2 output? I assume the client treats coordinates in its native coordinate space, not in the scale 2 coordinate space. I assume this is just done so that scale 2 surfaces work on scale 2 output (?), if so maybe the better way is to modify the coordinate_transform structure in surface.c instead. 

Apart from the coordinate transformations still being unclear and the need of get_scaled_output_size() I think we are on right track.

**Comment by ss1h2a3tw** [comment]
The reason of dividing boundary is that client use the coordinate of their surface,
when output scale=2, client is using the coordinate of resolution dividing the output scale.
So the pointer coordinate is half of output, when printing the cursor, we need to multiply by output scale.

And the input coordinate is right with my code.
img_20160623_175224
img_20160623_175333

weston-terminal scale=1
rhythmbox scale=2

**Comment by Cloudef** [comment]
Could the output scale taken into account in surface's coordinate_transform when it's attached to output here: https://github.com/Cloudef/wlc/blob/35caed59fa992a82774711622d86ac1172abd07b/src/resources/types/surface.c#L299 ?

**Comment by ss1h2a3tw** [comment]
OK I will try it later.

**Comment by Cloudef** [comment]
Also, probably good idea to include scale = 3 surface in your test too, this should in theory be downscaled to scale 2 output.

**Comment by ss1h2a3tw** [comment]
After my test,modify coordinate transform doesn't work.
Since
https://github.com/Cloudef/wlc/blob/309ddb03ee91b66ad4189d29ee651ac532d266a5/src/compositor/seat/pointer.c#L246
Subtracting offset first will make the offset been scaled by output scale too.

Since buffer_scale is informed by client,
I need to find some test client that I can define its buffer scale, or need some dirty hacking to test it.

**Comment by Cloudef** [comment]
Probably best way is to modify e.g. weston-simple-shm https://cgit.freedesktop.org/wayland/weston/tree/clients/simple-shm.c#n316, you may send custom scale with wl_surface_set_buffer_scale. I'll test this PR too to get sense how the coordinate transformations go, and that the visual coordinates match the logical ones, (even if client gets correct coordinates, it's easy to break coordinates compositor side, so hit detection, culling etc fails)

**Comment by ss1h2a3tw** [comment]
I found that simple-shm is using compositor version 1 so I need to change not only wl_surface_set_buffer_scale,
and I found simple-damage support scale as argument,
So I'm using it to test downscaling , seems working perfectly.

img_20160624_232212

**Comment by Cloudef** [comment]
Ok that seems quite promising. I'll pull the code and test too.

**Comment by ss1h2a3tw** [comment]
After my test of output scale = 3,4 upscale is working well as 2,
but because of the pointer coordinate is divide by output scale, it need move at least 3 pixels when output scale = 3, 4 pixels when output scale =4 and so on 

Maybe there's a better way for handling pointer coordinate.

**Comment by Cloudef** [comment]
I locally managed to do this with only changing output.c and surface.c, I'm killing some other bugs at the same time, I'll post some diffs for testing then.

**Comment by ss1h2a3tw** [comment]
I found that subsurface is rendering in the wrong position, I will fix it in few days.

**Comment by Cloudef** [comment]
I already fixed that locally, and also the fact subsurfaces were not scaled properly when buffer size is being stretched (due to minimum size). I'll try to get my code out this week for testing.

**Comment by ss1h2a3tw** [comment]
OK, Is there anything more I can help?

**Comment by Cloudef** [comment]
I'll push the code for testing when ready. The work here is great resource on what needs to be done.

**Comment by Cloudef** [comment]
https://github.com/Cloudef/wlc/tree/hidpi <- can you try this branch? For further fixes issues, probably good idea to open new PR against this.

**Comment by ss1h2a3tw** [comment]
I have test it, and I found two problems.
When spawning a client which support buffer scale at scale = 1,then change output scale to 2, but it doesn't set the buffer scale into 2.

Another problem is not quite related to scale, when using weston-subsurface to test , the pointer event seems not sending or sending wrong when the cursor is on subsurface.

I will add some screenshot later.

**Comment by Cloudef** [comment]
>When spawning a client which support buffer scale at scale = 1,then change output scale to 2, but it doesn't set the buffer scale into 2.

As far as wlc goes, it should send the new output information. It may be that the client just doesn't support this case.

>Another problem is not quite related to scale, when using weston-subsurface to test , the pointer event seems not sending or sending wrong when the cursor is on subsurface.

The pointer event problems may have been caused by unrelated change, that is basic support for "input area" of surfaces, but in sense it's incomplete as it only checks the area of views, not subsurfaces.

>The third problem is Xwayland seems not rendering in right scale.

I wonder if X11 has equivalent protocol to wayland scale that I should be aware of.. I really doubt it. Neverthless something to look into.

**Comment by Drakulix** [comment]
X11 has no similar protocol I know of, but many programs respect the "Xft.dpi" variable you may set in .Xresources. Additionally many GUI frameworks have custom environment variables.

GDK/GTK: `GDK_SCALE=2`
Qt: `QT_AUTO_SCREEN_SCALE_FACTOR=2`
Elementary: `ELM_SCALE=2`

I do not know, if wlc is supposed to set environment variables, but there should be at least a callback for changed scale factors, so an actual compositor might set those variables.

**Comment by Cloudef** [comment]
output.resolution callback already gets called, however it may be good idea to break the API of this callback and include from_scale and to_scale in the arguments as well. So you don't lose the information of old scale.

File: include/wlc/wlc.h (hunk 8 above)
Comment (SirCmpwn) [inline include/wlc/wlc.h]: :+1: for this function

**Comment by SirCmpwn** [comment]
Did some initial work on getting this branch working with sway. My branch is [here](https://github.com/SirCmpwn/sway/tree/hidpi). A few things I've noticed (some of this may be sway's fault and some of it may be wlc's fault, not sure yet):

* subsurfaces are 4x scaled if I use 2x scaling
* window borders are fucked up
* x windows are stretched rather than setting the xwayland DPI (this is a difficult one, since x only supports one DPI for the whole thing)

**Comment by SirCmpwn** [comment]
[Screenshot](https://sr.ht/Eh8T.png)

Use `output [name] scale 2` or similar to set the scale factor on sway.

**Comment by ss1h2a3tw** [comment]
@SirCmpwn  please use the branch https://github.com/Cloudef/wlc/tree/hidpi
my branch here have bug handling subsurface scale

**Comment by SirCmpwn** [comment]
Tried again with wlc's hidpi branch, works better. Pushed a new version to my hidpi branch.

**Comment by Cloudef** [comment]
Merged into master.
